Objects/listobject.c

```diff
@@ -3085,6 +3107,55 @@ static PySequenceMethods list_as_sequence = {
     list_inplace_repeat,                        /* sq_inplace_repeat */
 };
 
+static inline PyObject *
+list_slice_step_lock_held(PyListObject *a, Py_ssize_t start, Py_ssize_t step, Py_ssize_t len)
+{
+    PyListObject *np = (PyListObject *)list_new_prealloc(len);
+    if (np == NULL) {
+        return NULL;
+    }
+    size_t cur;
+    Py_ssize_t i;
+    PyObject **src = a->ob_item;
+    PyObject **dest = np->ob_item;
+    for (cur = start, i = 0; i < len;
+            cur += (size_t)step, i++) {
+        PyObject *v = src[cur];
+        dest[i] = Py_NewRef(v);
+    }
+    Py_SET_SIZE(np, len);
+    return (PyObject *)np;
+}
+
+static PyObject *
+list_slice_wrap(PyListObject *aa, Py_ssize_t start, Py_ssize_t stop, Py_ssize_t step)
+{
+    PyObject *res = NULL;
+    PyListObject *a = (PyListObject *)aa;
+    Py_BEGIN_CRITICAL_SECTION(a);
+    Py_ssize_t len = PySlice_AdjustIndices(Py_SIZE(a), &start, &stop, step);
+    if (len <= 0) {
+        res = PyList_New(0);
+    }
+    else if (step == 1) {
+        res = list_new_prealloc(len);
+        if (!res) {
+            return NULL;
+        }
+        PyObject **src = aa->ob_item;
+        PyObject **dest = ((PyListObject *)res)->ob_item;
+        for (Py_ssize_t cur = start, i = 0; i < len; cur += (size_t)step, i++) {
+            dest[i] = Py_NewRef(src[cur]);
+        }
+        Py_SET_SIZE(res, len);
```

Comment:
The `step == 1` case can be:

```
res = list_slice_lock_held(self, start, stop);
```

i.e., like the implementation in the code for this case before this PR:

https://github.com/python/cpython/blob/58c7919d05a360f3437a204960cddbeb78a71dce/Objects/listobject.c#L3118